Letterbook.Adapter.ActivityPub/Signatures/MastodonComponentBuilder.cs

```diff
@@ -0,0 +1,158 @@
+{
+    private readonly HttpRequestMessage _message;
+    private List<string> _derivedParams = new List<string>();
+    private List<string> _derivedParamsValues = new List<string>();
+    private List<string> _headerParams = new List<string>();
+    private List<string> _headerParamsValues = new List<string>();
+    
+    /* Draft 19 looks like
+     * "@signature-params": ("@method" "@authority" "@path" "content-digest" "content-length" "content-type");created=1618884473;keyid="test-key-rsa-pss"
+     */
+    
+    /* Mastodon/draft-cavage-8 looks like
+     * headers="(request-target) host date"
+     * Signature: keyId="https://my-example.com/actor#main-key",headers="(request-target) host date",signature="Y2FiYW...IxNGRiZDk4ZA=="
+     */
+    
+    public string SigningDocument => BuildSigningDocument();
+    public string? SigningDocumentSpec => BuildDocumentSpec();
+
+    public MastodonComponentBuilder(HttpRequestMessage message)
+    {
+        _message = message;
+    }
+
+    /// <summary>
+    /// Do nothing, this is not supported by draft-cavage-8
+    /// </summary>
+    /// <param name="component"></param>
+    public void Visit(SignatureComponent component)
+    { }
+    
+    public void Visit(HttpHeaderComponent httpHeader)
+    {
+        string fieldName = httpHeader.ComponentName;
+
+        if (TryGetHeaderValues(fieldName, out IEnumerable<string> values))
+        {
+            AddHeader(fieldName, String.Join(", ", values));
+        }
+        else
+        {
+            if (fieldName == "host")
+            {
+                AddHeader(fieldName, _message.GetDerivedComponentValue(SignatureComponent.Authority));
+            }
+        }
+    }
+    
+    /// <summary>
+    /// Do nothing, this is not supported by draft-cavage-8
+    /// </summary>
+    /// <param name="component"></param>
+    public void Visit(HttpHeaderDictionaryStructuredComponent httpHeaderDictionary)
+    {
+    }
+    
+    /// <summary>
+    /// Do nothing, this is not supported by draft-cavage-8
+    /// </summary>
+    /// <param name="component"></param>
+    public void Visit(HttpHeaderStructuredFieldComponent httpHeaderStructuredField)
+    {
+    }
+
+    /// <summary>
+    /// Derived components are not really a general case thing in draft-cavage-8. But, the request target is a required
+    /// component of the signature.
+    /// </summary>
+    /// <param name="derived"></param>
+    public void Visit(DerivedComponent derived)
+    {
+        var method = new DerivedComponent(DerivedComponents.Method);
+        switch (derived.ComponentName)
+        {
+            case DerivedComponents.RequestTarget:
+                AddRequestTarget($"{_message.GetDerivedComponentValue(method)} {_message.GetDerivedComponentValue(derived)}");
+                break;
+            case DerivedComponents.Authority:
+                AddHeader("host", _message.GetDerivedComponentValue(derived));
```

Comment:
You could factor `GetDerivedComponentValue(derived)` out of the switch, since it's used in both cases.